packages/engine/Source/Scene/GltfLoader.js

```diff
@@ -2394,6 +2395,75 @@ function loadInstanceFeaturesLegacy(
   }
 }
 
+/**
+ * Get an array of primitives for a given mesh. If the EXT_mesh_primitive_restart extension is present, use it to combine groups of primitives.
+ * If the extension is not present or its spec is violated, return the original mesh.primitives.
+ * @param {object} mesh A mesh from the glTF meshes array
+ * @returns {object[]} An array of mesh primitives
+ * @private
+ */
+function getMeshPrimitives(mesh) {
+  const meshExtensions = mesh.extensions ?? Frozen.EMPTY_OBJECT;
+  const primitiveRestartExtension = meshExtensions.EXT_mesh_primitive_restart;
+  const meshPrimitives = mesh.primitives;
+
+  if (!defined(primitiveRestartExtension)) {
+    return meshPrimitives;
+  }
+
+  // Note: per the spec, any violation of the extension's specification should cause us to fall back to mesh.primitives, if detecting the violation is feasible.
+
+  // Start with a copy of mesh.primitives. For each group, replace the first primitive in the group with a primitive representing the entire group,
+  // and set the rest of the primitives in the group to `undefined`.
+  // This allows us to identify which remaining primitives do not use primitive restart, and any errors involving a primitive appearing in more than one group.
+  const primitives = [...meshPrimitives];
```

Comment:
We had some issues with `push.apply` because it puts all elements on the stack (yeah, sounds crazy, but that's the way it is), and the same applies to the `[...spread]` operator (see https://github.com/CesiumGS/cesium/issues/12053 ). I don't know if we want to be prepared for meshes with 150000 primitives...

Reply:
Thanks for the `addAllToArray` suggestion, @javagl.  @eringram, once we add usage of that function to this PR, we should consider doing something similar to this parallel extension code in iTwin.js if necessary.

Reply:
Thanks, just updated this to use `addAllToArray`.